### DISCUSSION TITLE
Welcome to TorchSharp Discussions!

### DISCUSSION THREAD

**Discussion by dsyme** [xamarin/TorchSharp#333] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by NiklasGustafsson** [comment]
I'm Niklas Gustafsson. I've been at Microsoft for almost two decades, and worked on a number of developer-oriented technologies in Developer Division. 

I love .NET and I'm involved with this project because I want to make sure that .NET has the ML platform it deserves.

**Comment by dayo05** [comment]
Hi everyone!
I'm Dayo who lives on South Korea, 16years old(on 2021.11.09, today vs2022 has released!)
I'm not good at English but I think I can improve for other who wants use torch with C# :D
Thanks! (I'm newbie of contributing for opensource, so sometimes I can make mistake for it :( )

EDIT: I'm 17 years old now :D

**Comment by mfagerlund** [comment]
I'm Mattias Fagerlund, I've been programming since 1986, so I've been at it for a while. I'm a software developer, working as a consultant for clients.

I've been doing Machine Learning since 1996, I think, when I started playing around with Genetic Programming. I've since worked/experimented with a large number of different ML algorithms, lately mostly focusing on Deep Learning. I created a couple of my own Deep Learning frameworks with auto differentiation, but I decided it was time to use on one of the existing frameworks.

What I usually like to do is take a ML paper and implement it for myself - like an XGBoost clone I wrote, and then use those tools in stuff like Kaggle competitions, though it was a while ago since I competed, but I once had a ranking of #176 at kaggle https://www.kaggle.com/mfagerlund.

Pytorch seem to be going the strongest, and the TorchSharp integration has so far been very enjoyable to use.

**Comment by kaiidams** [comment]
Hello everyone!
I'm Katsuya Iida. I used to be a mobile app developer and lately been doing NLP/DL for 7 years. I am working on audio related things for my private projects https://github.com/kaiidams/voice100.

I like C# and believe that .NET has great potential for efficient deep learning and edge inference.

**Comment by TheBlackPlague** [comment]
Hello everyone.

I'm Shaheryar Sohail. Unlike most of you, I'm not as experienced in .NET (have barely 3-4 years experience). Please feel free to point out if I'm wrong about something — it helps me learn.

I've recently started StockNemo, a highly-optimized open-source chess engine,  which intends to use TorchSharp to train its evaluation neural network. For that, I would want this library to perform as fast as possible, and thus will try and help out as much as I can to make it happen.

I'm a believer that .NET code can run faster than C++. Thank you for reading!

**Comment by xhuan8** [comment]
Hi everyone,

I'm Xie Huan, I became a C#.NET programmer since 2006, currently I'm working on computer vision projects with C# and emguCV.

From 2019 I start to use Python and PyTorch on computer vision researches, such as 3D human pose and shape estimations, PyTorch is powerful and friendly to programmers. 

Now I'll try to implement deep learning models with TorchSharp, the first one is faster rcnn, it takes more time than I thought. It's glad that I can contribute to this project.

**Reply by NiklasGustafsson** [reply]
Glad to have you here as part of the TorchSharp community!

**Reply by GeorgeS2019** [reply]
I too very interested on 3D human pose and shape estimation. 
@xhuan8 @NiklasGustafsson @kaiidams 
With Torchscript, I wonder to what extent it is possible to take some of the PyTorch modules used in e.g. 3D human pose or generally other modules, and wrapped them using `torch.jit.compile` or through `torch.jit.save` for re-use in TorchSharp?

I have no experience with `torch.jit.compile` versus `torch.jit.save` for bringing PyTorch codes to TorchSharp without porting python to .NET

**Reply by saint4eva** [reply]
I hope there will not be python-styled codes  scattered everywhere in C# codebase?

**Reply by xhuan8** [reply]
> I hope there will not be python-styled codes scattered everywhere in C# codebase?

There are many benefits if you convert some Python project to C#. Most of existing code are written by Python, so keep the Python coding style will help saving a lot of time.

**Reply by dayo05** [reply]
> > I hope there will not be python-styled codes scattered everywhere in C# codebase?
> 
> There are many benefits if you convert some Python project to C#. Most of existing code are written by Python, so keep the Python coding style will help saving a lot of time.

Both case have benefits, C# styled codebase will good for beginning with C#. But, Python styled codebase will good for converting Python to C#. Somedays, maybe we should add extra wrapper for C# styled codebase.

**Reply by saint4eva** [reply]
> > I hope there will not be python-styled codes scattered everywhere in C# codebase?
> 
> There are many benefits if you convert some Python project to C#. Most of existing code are written by Python, so keep the Python coding style will help saving a lot of time.

I do not think so, because the contributing developers are not python developers. There is a reason for code-standard and styles. Every ecosystem becomes productive when they are familiar with the code they read - so contributing becomes easier.

**Reply by NiklasGustafsson** [reply]
We made the decision about a year ago to follow the initiative of the SciSharp community and prioritize the ability to lower the cost of transferring Python examples, tutorials, etc. when using TorchSharp. This was a tough decision to make, and we were fully aware of the controversial nature of this decision.

The principle is that for any APIs that are a direct "copy" of the PyTorch APIs, we use the conventions used in PT. For everything else, i.e. things that are introduced for TorchSharp and is unique to it, we use .NET conventions. See, for example, the NewDisposeScope APIs.

We recognize that this is not for everyone, but it's the decision we made. TorchSharp is meant to be a lower-level API, and we intend to adhere to .NET naming conventions when integrating it into ML.NET, for example. In the long run, we expect .NET developers to use  higher-level APIs with tooling support (i.e. ML.NET).

**Reply by GeorgeS2019** [reply]
One year ago, the TorchSharp and many projects in SciSharp were still far behind the wrapping coverage of pyhton API (e.g. Tensorflow.NET etc.). It was always unclear what options are available for c# (or .NET) developers when come to Deep AI.  Especially moving forwards.

One year ago, ML.NET conducted a survey and there was an overwhelming need/demand for Deep AI for .NET communities. 

Microsoft released Deep AI (CNTK) back in 2016, the adoption by .NET users was insufficient, although it was designed from the ground up to address the Deep AI of .NET communities. => [Has Microsoft abandoned CNTK?](https://stackoverflow.com/questions/55831498/has-microsoft-abandoned-cntk)

It was prudent a year back that this highly controversial decision (based on PyTorch-Look-alike syntax) was taken to EASE the adoption of PyTorch for .NET communities through TorchSharp.

This is not based on Python.NET but the same strategy as adopted by e.g. RUST community by wrapping around LibTorch. 

I am a strong believer that this highly controversial decision, together with the positive attitude of many contributors who understand the urgent need to have Deep AI option for .NET developers, joined and contributed to a more promising future for Deep AI for .NET developers. 

NOW, it is no longer doubt if .NET langauages (e.g. C#) could be part of the fast-changing Deep AI communities. 

This brings hope that as a .NET developer, there is NOW a concrete OPTION to engage in Deep AI. 

With this HOPE, it is amazing to see that there are now more commitments/contributions to speed up the task to complete the coverage of PyTorch API but in .NET.

In my view, the controversial decision is KEY to growing the confidence among .NET communities.

Once there is an active and vibrant Deep AI .NET communities, once the github is full of many different .NET deep AI projects, the sky is the limit for emerging .Deep AI that adheres 100% to all .NET syntax and standards. 

Once that has happened, TorchSharp has served the transition needed to give HOPE to .NET users who want to do Deep AI but stay with .NET technology.

**Comment by mfagerlund** [comment]
Hard thumbs up for you 👍 and the rest of the team making this happen 👍

Den tors 27 okt. 2022 kl 20:24 skrev GeorgeS2019 ***@***.***>:

> One year ago, the TorchSharp and many projects in SciSharp were still far
> behind the wrapping coverage of pyhton API (e.g. Tensorflow.NET etc.). It
> was always unclear what options are available for c# (or .NET) developers
> when come to Deep AI. Especially moving forwards.
>
> One year ago, ML.NET conducted a survey and there was an overwhelming
> need/demand for Deep AI for .NET communities.
>
> Microsoft released Deep AI (CNTK) back in 2016, the adoption by .NET users
> was insufficient, although it was designed from the ground up to address
> the Deep AI of .NET communities.
>
> It was prudent a year back that this highly controversial decision (based
> on PyTorch-Look-alike syntax) was taken to EASE the adoption of PyTorch for
> .NET communities through TorchSharp.
>
> This is not based on Python.NET but the same strategy as adopted by e.g.
> RUST community by wrapping around LibTorch.
>
> I am a strong believer that the highly controversial decision, together
> with many contributors who see the urgent need for Deep AI option for .NET
> developers, contributed a more promising future for Deep AI for .NET
> developers.
>
> NOW, it is no longer doubt if .NET langauages (e.g. C#) could be part of
> the fast-changing Deep AI communities.
>
> This brings hope that as a .NET developer, there is NOW a concrete OPTION
> to engage in Deep AI.
>
> With this HOPE, it is amazing to see that there are now more
> commitments/contributions to speed up the task to complete the coverage of
> PyTorch API but in .NET.
>
> In my view, the controversial decision is KEY to growing the confidence
> among .NET communities.
>
> Once there is an active and vibrant Deep AI .NET communities, once the
> github is full of many different .NET deep AI projects, the sky is the
> limit for emerging .Deep AI that adheres 100% to all .NET syntax and
> standards.
>
> Once that has happened, TorchSharp has served the transition needed to
> give HOPE to .NET users who want to do Deep AI but stay with .NET
> technology.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/discussions/333#discussioncomment-3984880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556N2RHJCSVMVBWUEDGLWFLCFTANCNFSM5DDU6IKQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
/Mattias Fagerlund

**Comment by FancyFurret** [comment]
Hello! I've been programming for about 10 years in many different languages, my favorites being C#, C++, and Typescript. Currently working as a software developer writing backend business stuff, but in my free time I love learning about new things and writing interesting apps. Recently I started learning pytorch for a few ideas I had to improve the workflow of drawing sprite animations, but am switching over to TorchSharp since C# is so much better.

**Reply by saint4eva** [reply]
That is good. Kindly share your experiences as you go along. 👍🏽

**Comment by ChengYen-Tang** [comment]
Hello everyone,
I'm Kenneth Tang living in Taiwan.
I started using VB.net in high school and turned into a C# developer in college. Now I'm a reinforcement learning researcher.
I have a side project where I spend my off hours trying to implement a stable baselines3-like reinforcement learning framework in the C# ecosystem. But there are many bottlenecks in language features, such as python has **kwargs, so it is progressing slowly.

I think the biggest bottleneck is that python has [pickle](https://docs.python.org/3/library/pickle.html), which can simply save the state of all objects, including objects from other libraries. But C# has no perfect solution

**Reply by xhuan8** [reply]
currently I use dictionary to handle **kwargs, requires a lot of work.

**Reply by ChengYen-Tang** [reply]
However, other libraries do not support this method.
Do you have any suggested method for storing object state?

**Reply by xhuan8** [reply]
as metioned in https://github.com/dotnet/TorchSharp/blob/main/docfx/articles/saveload.md
there are some python scripts can convert between TorchSharp and PyTorch, but I haven't tested yet. 

https://github.com/dotnet/TorchSharp/blob/main/src/Python/importsd.py
https://github.com/dotnet/TorchSharp/blob/main/src/Python/exportsd.py

**Reply by ChengYen-Tang** [reply]
What I mean is that SB3 uses pickle to save the state of the environment. It doesn't just save weights, it can be numpy ndarray, shape or any object state of pytorch, openai gym.

**Reply by NiklasGustafsson** [reply]
Yes, pickling is not something we're going to be handling within TorchSharp. I think there are some external attempts at implementing Python pickling for .NET, but I'm not sure about the quality or capabilities, so I won't make any recommendations. There are various .NET serialization libraries, of course, but that's different. 

The scripts that @xhuan8 mention above allow you to share _weights_ between PyTorch and TorchSharp, and it builds on the ability to share tensors. However, that's pretty much it as far as state sharing between PT and TS goes. Well, there's TorchScript, of course.

**Reply by NiklasGustafsson** [reply]
>> But there are many bottlenecks in language features, such as python has **kwargs, so it is progressing slowly

What other language bottlenecks are there that are getting in the way? It'd be great to know and see how we can help work around them.

**Reply by ChengYen-Tang** [reply]
> Yes, pickling is not something we're going to be handling within TorchSharp. I think there are some external attempts at implementing Python pickling for .NET, but I'm not sure about the quality or capabilities, so I won't make any recommendations. There are various .NET serialization libraries, of course, but that's different.
> 
> The scripts that @xhuan8 mention above allow you to share _weights_ between PyTorch and TorchSharp, and it builds on the ability to share tensors. However, that's pretty much it as far as state sharing between PT and TS goes. Well, there's TorchScript, of course.

I have spent some time researching dotnet's serialization library, but have a few questions. Some libraries must pass parameters when creating an instance, and these objects usually cannot be deserialized.
BinaryFormatter can avoid this problem, but BinaryFormatter must mark [Serializable] in Class to serialize objects.

I think to achieve a function similar to pickle, it seems that the support of the dotnet team is required

**Reply by ChengYen-Tang** [reply]
> > > But there are many bottlenecks in language features, such as python has **kwargs, so it is progressing slowly
> 
> What other language bottlenecks are there that are getting in the way? It'd be great to know and see how we can help work around them.

Python can initialize subcategories before initializing parent categories, or parent categories when subcategories are halfway initialized, which is annoying.
```
def __init__(self):
    # do something
    super().__init__()
    # do something
```

I have developed a library that can do kwargs similar to python, although it is still a bit awkward to use in the strongly typed.
https://github.com/ChengYen-Tang/Kwargs.Net
Another is that the library is Union, which allows us to return different types like python.
https://github.com/ChengYen-Tang/Typing.Union

**Comment by GilesBathgate** [comment]
Hi I am @GilesBathgate I've made some small contributions recently. My background is that I have been a software developer for an Robotic Process Automation (RPA) company using .NET for the last 20 years. :older_man:  I'm a relative newbie to ML, but learning fast. I also have experience in the open source C++ library CGAL, which I have contributed to, because I use it in my 3D printing project called [RapCAD](https://rapcad.org/)

**Reply by ChengYen-Tang** [reply]
Hi @GilesBathgate,
I'm curious how machine learning is used in RapCAD

**Reply by GilesBathgate** [reply]
@ChengYen-Tang It isn't. However, leaning how linear algebra works is a good foundation for geometry processing, and ML.

**Reply by NiklasGustafsson** [reply]
I think it's very interesting to consider uses other than ML for a library like TorchSharp. After all, it's a fairly general numeric library based on N-dimensional vectors. The fact that it also supports back-propagation isn't necessarily something you have to take advantage of.

**Reply by ChengYen-Tang** [reply]
Honestly, I prefer to call Torch and Tensorflow a math framework rather than a machine learning framework, and I use them a lot to solve math problems.
These frameworks lower the threshold for me to use GPU to solve math problems, I don't need to write troublesome CUDA code🤣

**Reply by NiklasGustafsson** [reply]
> Honestly, I prefer to call Torch and Tensorflow a math framework rather than a machine learning framework, and I use them a lot to solve math problems. These frameworks lower the threshold for me to use GPU to solve math problems, I don't need to write troublesome CUDA code🤣

Yes, specifically linear algebra -- it's cumbersome and inefficient to use for non-vector-based math.

**Comment by yueyinqiu** [comment]
Hello everyone! I'm a college student in China, majoring in physics currently and deciding to transfer to software engineering during my doctoral studies. So actually I know little about deep learning, but I suppose I am quite familiar with C#. I'm here because Python is really awful, in my opinion from almost any perspective, and I hope my future works about deep learning could be based on C#.

**Reply by GeorgeS2019** [reply]
@yueyinqiu 
It is important that you have shared that you have physics background

The data science and the whole industries, driven by the vision of Nvidia, towards 3D visualisation and real world simulation.

Cloud providers, Microsoft, AWS, and Google are still lacking consistent strategy to compete with Nvidia 

Learn Godot with TorchSharp

This is how you can combine your physics with C#

**Reply by GeorgeS2019** [reply]
https://github.com/asieradzk/RL_Matrix

Godot Jolt physics simulation using TorchSharp 
https://github.com/smtmRadu/GodotAI

**Comment by shaltielshmid** [comment]
I'm Shaltiel Shmidman. I'm the lead developer of the deep learning lab at [Dicta](https://dicta.org.il), where we're dedicated to advancing AI and ML technologies, specifically for Hebrew applications.

Like many of you, I love .NET and want to able to deploy our deep learning products using the robust ASP.NET environment. 6 years ago, before TorchSharp was fully developed, I ported the [DyNet](https://github.com/clab/dynet) deep learning library to C# myself, and we currently serve our products using that library ([DyNetSharp](https://www.nuget.org/packages/DynetSharp)). 

As time has progressed, DyNet has faded out of support, and I am thrilled that TorchSharp has evolved. I am actively contributing to this project and am eager to help provide .NET with a stable and powerful ML platform that complements the other advanced features being constantly introduced.

**Comment by asieradzk** [comment]
Hey Everyone, I am Adrian. Chemistry graduate, learning software engineering since 2019. 

Right now I am working on my PhD in Chemical Engineering at ETH Zurich where I'm using deep reinforcement learning to design microfluidic devices. To that end I am creating [RLMatrix](https://github.com/asieradzk/RL_Matrix) and I am fully intent on replacing it as Deep Reinforcement Learning toolkit for game engines. (I'm a big hater of Unity's ML Agents). I also want to eclipse Rlib with C#.

Shout-out to fellow science major @yueyinqiu

**Comment by mfagerlund** [comment]
Have you been able to use TorchSharp in a deployed unity project?

Den ons 26 juni 2024 kl 15:45 skrev Adrian Sieradzki <
***@***.***>:

> Hey Everyone, I am Adrian. Chemistry graduate, learning software
> engineering since 2019.
>
> Right now I am working on my PhD in Chemical Engineering at ETH Zurich
> where I'm using deep reinforcement learning to design microfluidic devices.
> To that end I am creating RLMatrix and I am fully intent on replacing it as
> Deep Reinforcement Learning toolkit for game engines. (I'm a big hater of
> Unity's ML Agents). I also want to eclipse Rlib with C#.
>
> Shout-out to fellow science major @yueyinqiu
> <https://github.com/yueyinqiu>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/discussions/333#discussioncomment-9883102>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556PS4DYU62V42D7NPUTZJLAW3AVCNFSM5DDU6IK2U5DIOJSWCZC7NNSXTOKENFZWG5LTONUW63SDN5WW2ZLOOQ5TSOBYGMYTAMQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
/Mattias Fagerlund

**Reply by asieradzk** [reply]
> Have you been able to use TorchSharp in a deployed unity project? Den ons 26 juni 2024 kl 15:45 skrev Adrian Sieradzki < 

I just am just finishing conditional compilation to .net standrd 2.0 so maybe over the weekend/next weel :) Think I'll look for a way to replace it as backed with Unity ML agents just to flex on Unity. 

You can use it with Unity right now if you host RLMatrix server asp.net core application and use the Remote client inside your Unity project. But thats over socket, shared memory mode will have to wait.


(For the uninitiated there is no way to use DRL in game engines with shared memory mode, solutions like ML Agents all use socket to talk to python server)

**Comment by mfagerlund** [comment]
Nice, let us know how it goes!

Den fre 28 juni 2024 kl 12:45 skrev Adrian Sieradzki <
***@***.***>:

> Have you been able to use TorchSharp in a deployed unity project? Den ons
> 26 juni 2024 kl 15:45 skrev Adrian Sieradzki <
>
> I just am just finishing conditional compilation to .net standrd 2.0 so
> maybe over the weekend/next weel :) Think I'll look for a way to replace it
> as backed with Unity ML agents just to flex on Unity.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/discussions/333#discussioncomment-9903669>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556OCHAFZNPSHGSTQUBLZJU5EZAVCNFSM5DDU6IK2U5DIOJSWCZC7NNSXTOKENFZWG5LTONUW63SDN5WW2ZLOOQ5TSOJQGM3DMOI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
/Mattias Fagerlund

**Reply by asieradzk** [reply]
I managed to get it to work in Unity but its underwhelming and disappointing:
-First of all dealing with Unity's bizarre main thread was (and still is) a lot of pain. Even on newer version where "Awaitable" exists.
-Running RLMatrix directly inside Unity does work and is pretty fast but is still a lot slower than running server version on the same system and communicating over local socket. I've no idea what's causing this massive hit in performance but its not something I've seen when profiling a raw .net core 3.1 application - I went through the trouble of adding multiple compilation targets to RLMatrix, tested it and its rubbish inside Unity despite everything.

**Comment by mfagerlund** [comment]
Sorry to hear that. I've had performance problems in Unity with code that
runs very quickly outside unity and I was unable to find any usable fix for
my issues.

/m

Den mån 12 aug. 2024 kl 14:21 skrev Adrian Sieradzki <
***@***.***>:

> I managed to get it to work in Unity but its underwhelming and
> disappointing:
> -First of all dealing with Unity's bizarre main thread was (and still is)
> a lot of pain. Even on newer version where "Awaitable" exists.
> -Running RLMatrix directly inside Unity does work and is pretty fast but
> is still a lot slower than running server version on the same system and
> communicating over local socket. I've no idea what's causing this massive
> hit in performance but its not something I've seen when profiling a raw
> .net core 3.1 application - I went through the trouble of adding multiple
> compilation targets to RLMatrix, tested it and its rubbish inside Unity
> despite everything.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/TorchSharp/discussions/333#discussioncomment-10313763>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADD556PNK2RE6YZEPNGRFI3ZRCSGFAVCNFSM5DDU6IK2U5DIOJSWCZC7NNSXTOSENFZWG5LTONUW63SDN5WW2ZLOOQ5TCMBTGEZTONRT>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
/Mattias Fagerlund

**Reply by asieradzk** [reply]
Must be something wrong with MONO that unity is using, because same as you I've noticed perfect performance in vanilla .netstandard 2.0 application, but Unity tanks.

Moreso when I run TorchSharp code in asp.net core server and still use it in blocking manner inside Unity - its still faster than running it directly inside.... lol It's same, identical code i'm just moving torcsharp calls to a separate application.

**Comment by NiklasGustafsson** [comment]
I just wanted to let everyone here know that I'm leaving TorchSharp to rejoin the C++ team here at Microsoft. TorchSharp will now be in the capable hands of developers from our office in Prague, the same team responsible for our work on libtorch on WIndows. 

Note my use of plural 'developers' -- TorchSharp will now be supported by more than one person. That's just pure goodness.

It's been a wonderful 4.5 years for me. Thanks to all of you who have contributed your time and enthusiasm!

To the team in Prague -- please introduce yourselves in this thread with a short backgrounder.

**Reply by GeorgeS2019** [reply]
@NiklasGustafsson 

Thank you for stepping forwards to support pytorch like codes when many were against that 4.5 years ago.

Look forwards to getting to know the new team members!

With a long ToDo wishes ☺️

**Reply by alinpahontu2912** [reply]
Hello everyone, I'm Alin, one of the developers stepping in to continue Niklas' work on TorchSharp. Over the past year and a half, I've been focused on bringing the PyTorch ecosystem to Windows on Arm64. I'm excited to join the effort on TorchSharp and collaborate with you—the community developers—to build this project together!

**Reply by GeorgeS2019** [reply]
@alinpahontu2912 

Hi Stefan, I was with @NiklasGustafsson right from the beginning.  Over the years, because we only have one official ( actually part-time) support, we could not explore more time-challenging topics we have discussed over the years. 

Torchsharp is now actively used to advance ML.NET and AI Agents using .NET. 

It would be of interest to us that we have some ideas what are yours as well as your other colleague's priorities for TorchSharp. 

So the TorchSharp community could provide relevant feedback to push torchsharp development to one that further our original vision: to keep the .NET developers doing Deep Learning to stay within the .NET frameworks without needing venture into pytorch or python codes.

**Reply by ghost** [reply]
Thanks @NiklasGustafsson for all the efforts so far.

I'm Ozan, and as Niklas mentioned, I'm working from the [Microsoft Prague](https://news.microsoft.com/europe/features/microsoft-prague/) (aka MDCP) office.

As a team, we also sometimes contribute to some of the **PyTorch - Windows** topics. Over time, I hope to further improve the `TorchSharp` repository/pipelines with new updates and keep it synced with the latest features. On the other hand, community contributions/issues/PRs are still very important and highly appreciated.

Again, big thanks to @NiklasGustafsson and to all the contributors of the repository.

**Reply by iremyux** [reply]
Thanks @NiklasGustafsson for your efforts and good luck on your new team :) 

Hi everyone, I'm Irem. I've been working on Pytorch on Windows x64 & Arm64 for more than 2 years. 
I'm thrilled to be part of this project and collaborate with the community. I look forward to achieving great things together.

**Reply by GeorgeS2019** [reply]
@NiklasGustafsson 

Do you have goals which would not be possible in the past, but now seems possible as we now have official 3 persons supporting this project?

**Reply by dsyme** [reply]
Great to see TorchSharp continuing!

**Reply by GeorgeS2019** [reply]
@dsyme @NiklasGustafsson @LittleLittleCloud 

Recently I attended a talk on how the extensive bias of python in e.g. AI Agent could potentially means the eventual inevitable death of other languages 

This is why Microsoft need to commit more in TorchSharp..the critical decision of using python like syntax back in 2021 could mean the survival of c# language a bit longer time due to the existence of TorchSharp.

It is not enough to have just ML.net integrated with TorchSharp, but investing effort to port , through generative AI code generation, many future proof pytorch libraries to TorchSharp. 

E.g. SB3 to TorchSharp.

Currently there is a small effort from Microsoft on RL, but it only covers a small subset of RL.

Microsoft needs to start thinking long term to protect c#, but being decisive and commit to preserve c# and .Net ecosystem through TorchSharp 

The 3 new guys from Prague, they need to understand this vision..how to coordinate with ML.net and c# side of Autogen.

The inconvenient decision we made back in June 2021 to choose pytorch Syntax for TorchSharp, history would one day judge it as critical moment to preserve the .Net ecosystem

**Reply by GeorgeS2019** [reply]
@whuanle

Soon, Microsoft will realise..it is important to have TorchSharp as complete code coverage as pytorch as possible 

Keep doing what U are doing , keep us update.. will get Microsoft aware of your tutorials,

Ideally, the codes need to be as semantics similar to the pytorch codes of Microsoft tutorials as possible ...to convince Microsoft 

https://github.com/dotnet/TorchSharp/issues/1450

**Comment by GeorgeS2019** [comment]
@LittleLittleCloud

I need your feedback, if you see there are areas where you would like to advance TorchSharp to better support ML.NET and Geneartive AI

**Reply by LittleLittleCloud** [reply]
#1 Provide built-in support for int8 and int4 quantization would be very helpful for improving LLM inference for models that implement on top of torchsharp.

**Reply by asieradzk** [reply]
Ditto, deep reinforcement learning would also benefit immensely from quantization.
